### PR TITLE
[SPARK-50212][BUILD][3.5] Fix the conditional check for executing the `build` task in `build_and_test.yml`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
Comparing https://github.com/apache/spark/blob/c53dac05058c48ae1edad7912e8cc82533839ca0/.github/workflows/build_and_test.yml#L102 and https://github.com/apache/spark/blob/9d472661daad4703628e9fbf0ba9922abeed7354/.github/workflows/build_and_test.yml#L97, the master branch uses `dev/is-changed.py` to check for changes in additional modules: `variant`, `api`, `streaming-kinesis-asl`, `protobuf`, and `connect`. Among these, the `api`, `protobuf`, and `connect` modules also exist in the `branch-3.5` and should be checked as well.

Therefore, this pr includes the following changes:

1. Adds `is-changed` checks for the `api`, `protobuf`, and `connect` modules.
2. In `dev/sparktestsupport/modules.py`, adds the definition for the `api` module, aligning with the master branch.
3. In `dev/sparktestsupport/modules.py`, adds the definition for the `utils` module, aligning with the master branch. Prior to this PR, although `dev/is-changed.py` was used to check for changes in the `utils` module, its definition was missing from `dev/sparktestsupport/modules.py`.

### Why are the changes needed?
Fix the conditional check for executing the `build` task in `build_and_test.yml`

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Pass GitHub Actions

- Manually verified the effectiveness of this pull request:

Before:

Set the latest commit by executing `export APACHE_SPARK_REF=9d472661daad4703628e9fbf0ba9922abeed7354`. Then manually edit files in the `api`, `protobuf`, or `connect` modules and commit them. Run the following command:

```
./dev/is-changed.py -m "core,unsafe,kvstore,avro,utils,network-common,network-shuffle,repl,launcher,examples,sketch,graphx,catalyst,hive-thriftserver,streaming,sql-kafka-0-10,streaming-kafka-0-10,mllib-local,mllib,yarn,mesos,kubernetes,hadoop-cloud,spark-ganglia-lgpl,sql,hive,connect,protobuf,api"
```

The console will print `false`.

After:

Set the latest commit by executing `export APACHE_SPARK_REF=41446b3d98cfccf5c6f6ddb8bc3c7c6c1b1c3f54`. Then manually edit files in the `api`, `protobuf`, or `connect` modules and commit them. Run the same command as before:

```
./dev/is-changed.py -m "core,unsafe,kvstore,avro,utils,network-common,network-shuffle,repl,launcher,examples,sketch,graphx,catalyst,hive-thriftserver,streaming,sql-kafka-0-10,streaming-kafka-0-10,mllib-local,mllib,yarn,mesos,kubernetes,hadoop-cloud,spark-ganglia-lgpl,sql,hive,connect,protobuf,api"
```

The console will now print `true`.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by LuciferYang** [comment]
cc @dongjoon-hyun

File: dev/sparktestsupport/modules.py [hunk 1]
```diff
@@ -188,6 +196,14 @@ def __hash__(self):
     ],
 )
 
+api = Module(
+    name="api",
+    dependencies=[utils, unsafe],
+    source_file_regexes=[
+        "sql/api/",
```
Comment (dongjoon-hyun) [inline dev/sparktestsupport/modules.py:203]: So, is this missed since Apache Spark 3.5.0 until now?

**Review by dongjoon-hyun — APPROVED** [review]
The proposed PR looks correct to me. Thank you, @LuciferYang .

**Comment by dongjoon-hyun** [comment]
Feel free to merge when the PR passed the CI, @LuciferYang .

File: dev/sparktestsupport/modules.py [hunk 2]
```diff
@@ -113,6 +113,14 @@ def __hash__(self):
     ],
 )
 
+utils = Module(
+    name="utils",
+    dependencies=[tags],
+    source_file_regexes=[
+        "common/utils/",
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply dev/sparktestsupport/modules.py:120]: The first time `utils` and `api` were added to `modules.py` was in 2023.10.25: 

https://github.com/apache/spark/pull/43501

but they were not backported to branch-3.5.

File: dev/sparktestsupport/modules.py (hunk 1 above)
Comment (LuciferYang) [reply dev/sparktestsupport/modules.py:203]: Yes ...

Comment (LuciferYang) [reply dev/sparktestsupport/modules.py:203]: Yes ...

**Comment by LuciferYang** [comment]
[5d60a74](https://github.com/apache/spark/pull/48744/commits/5d60a749180cf2d647e6b8bd08b86e2b9bd906e3)  is refer to https://github.com/apache/spark/pull/43501 and https://github.com/apache/spark/pull/43818/files

**Comment by LuciferYang** [comment]
> Feel free to merge when the PR passed the CI, @LuciferYang .

OK

**Comment by LuciferYang** [comment]
Merged into branch-3.5. Thanks @dongjoon-hyun ~

**Pull request closed without merging**